endpoints/zones/src/main.cpp

```diff
@@ -248,58 +250,94 @@ namespace
+					// clang-format off
+					res.body() = json{
+						{"irods_response", {
+							{"status_code", e.code()},
+							{"status_message", e.client_display_what()}
+						}}
+					}.dump();
+					// clang-format on
+				}
+				catch (const std::exception& e) {
+					logging::error("{}: {}", fn, e.what());
+					res.result(http::status::internal_server_error);
+				}
 
-						const auto user = acl_user_pair.substr(colon_pos + 1);
+				res.prepare_payload();
 
-						auto acl = adm::zone_collection_acl::null;
-						if (acl_sv == "read") {
-							acl = adm::zone_collection_acl::read;
-						}
-						else if (acl_sv != "null") {
-							logging::error(
-								"{}: Invalid value for [value] parameter. Received [{}]. ACL must be [read] or [null].",
-								fn,
-								acl_user_pair);
-							return _sess_ptr->send(irods::http::fail(res, http::status::bad_request));
-						}
+				return _sess_ptr->send(std::move(res));
+			});
+	} // op_modify
 
-						adm::client::modify_zone(
-							conn, name_iter->second, adm::zone_collection_acl_property{acl, std::string{user}});
+	IRODS_HTTP_API_ENDPOINT_OPERATION_SIGNATURE(op_set_zone_collection_permission)
+	{
+		auto result = irods::http::resolve_client_identity(_req);
+		if (result.response) {
+			return _sess_ptr->send(std::move(*result.response));
+		}
+
+		const auto client_info = result.client_info;
+
+		irods::http::globals::background_task(
+			[fn = __func__, client_info, _sess_ptr, _req = std::move(_req), _args = std::move(_args)] {
+				logging::info("{}: client_info.username = [{}]", fn, client_info.username);
+
+				http::response<http::string_body> res{http::status::ok, _req.version()};
+				res.set(http::field::server, irods::http::version::server_name);
+				res.set(http::field::content_type, "application/json");
+				res.keep_alive(_req.keep_alive());
+
+				try {
+					const auto name_iter = _args.find("name");
+					if (name_iter == std::end(_args)) {
+						logging::error("{}: Missing [name] parameter.", fn);
+						return _sess_ptr->send(irods::http::fail(res, http::status::bad_request));
 					}
-					else {
-						logging::error("{}: Invalid value for [property] parameter.", fn);
+
+					const auto perm_iter = _args.find("permission");
+					if (perm_iter == std::end(_args)) {
+						logging::error("{}: Missing [permission] parameter.", fn);
+						return _sess_ptr->send(irods::http::fail(res, http::status::bad_request));
+					}
+
+					const auto user_iter = _args.find("user");
+					if (user_iter == std::end(_args)) {
+						logging::error("{}: Missing [user] parameter.", fn);
+						return _sess_ptr->send(irods::http::fail(res, http::status::bad_request));
+					}
+
+#if 0
+					// TODO Are zones supported?
+					const auto user_zone_iter = _args.find("user_zone");
+					if (value_iter == std::end(_args)) {
+						logging::error("{}: Missing [user] parameter.", fn);
 						return _sess_ptr->send(irods::http::fail(res, http::status::bad_request));
 					}
+#endif
```

Comment:
Guessing this is an ongoing investigation?

Reply:
Correct.

Reply:
See https://github.com/irods/irods_client_http_api/pull/282#discussion_r1669318629.